### PR TITLE
planner: consider disk cost in hashJoin

### PR DESCRIPTION
**Pull request by SunRunAway**
### What problem does this PR solve? 

part of #11607

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -390,17 +392,38 @@ func (p *PhysicalIndexJoin) GetCost(outerTask, innerTask task) float64 {
...
+}
+
 // GetCost computes cost of hash join operator itself.
 func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
 	innerCnt, outerCnt := lCnt, rCnt
+	inner := p.children[0]
 	// Taking the right as the inner for right join or using the outer to build a hash table.
 	if (p.InnerChildIdx == 1 && !p.UseOuterToBuild) || (p.InnerChildIdx == 0 && p.UseOuterToBuild) {
 		innerCnt, outerCnt = rCnt, lCnt
+		inner = p.children[1]
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply planner/core/task.go]: Renamed.

File: planner/core/task.go [hunk 2]
```diff
@@ -431,20 +454,36 @@ func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
...
+			cpuCost += buildCnt * sessVars.CPUFactor
+		} else {
+			cpuCost += buildCnt * sessVars.CPUFactor / float64(p.Concurrency)
+		}
+		diskCost += buildCnt * sessVars.DiskFactor * rowSize
 	}
-	return cpuCost + memoryCost
+
+	if spill {
+		memoryCost *= float64(memQuota) / (rowSize * buildCnt)
```
Comment (XuHuaiyu) [inline planner/core/task.go]: What does this mean?

Comment (SunRunAway) [reply planner/core/task.go]: Does these code can describe the meaning?
```go
memUsed := float64(memQuota)
originMemUsed := rowSize * buildCnt
factor := memUsed / originMemUsed
memoryCost *= factor
```

Comment (XuHuaiyu) [reply planner/core/task.go]: Got you, I misunderstood it before.
I think `memoryCost *= float64(memQuota) / (rowSize * buildCnt)` is ok.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @eurekaka

File: planner/core/task.go [hunk 3]
```diff
@@ -431,20 +454,36 @@ func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
 	// Cost of quering hash table is cheap actually, so we just compute the cost of
```
Comment (eurekaka) [inline planner/core/task.go]:
```suggestion
	// Cost of querying hash table is cheap actually, so we just compute the cost of
```
Fix this typo by hand in this PR?

File: planner/core/task.go [hunk 4]
```diff
@@ -431,20 +454,36 @@ func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
...
 	probeCost := numPairs * sessVars.CPUFactor
+	probeDiskCost := numPairs * sessVars.DiskFactor * rowSize
 	// Cost of evaluating outer filter.
 	if len(p.LeftConditions)+len(p.RightConditions) > 0 {
 		// Input outer count for the above compution should be adjusted by selectionFactor.
 		probeCost *= selectionFactor
-		probeCost += outerCnt * sessVars.CPUFactor
+		probeCost += probeCnt * sessVars.CPUFactor
+		probeDiskCost *= selectionFactor
+		probeDiskCost += probeCnt * sessVars.DiskFactor * rowSize
```
Comment (eurekaka) [inline planner/core/task.go]: What is this line for?

Comment (SunRunAway) [reply planner/core/task.go]: Misunderstood, I'll remove it.

**Comment by SunRunAway** [comment]
@eurekaka 
All comments are addressed, PTAL again, thanks.

File: statistics/table.go [hunk 5]
```diff
@@ -694,6 +695,28 @@ func (coll *HistColl) GetAvgRowSize(cols []*expression.Column, isEncodedKey bool
 	return size + float64(len(cols))
 }
 
+// GetAvgRowSizeListInDisk computes average row size for given columns.
+func (coll *HistColl) GetAvgRowSizeListInDisk(cols []*expression.Column, padChar bool) (size float64) {
+	if coll.Pseudo || len(coll.Columns) == 0 || coll.Count == 0 {
+		for _, col := range cols {
+			size += float64(chunk.EstimateTypeWidth(padChar, col.GetType()))
```
Comment (eurekaka) [inline statistics/table.go]: Change `pseudoColSize` in `GetAvgRowSize` to `EstimateTypeWidth` as well?

Comment (SunRunAway) [reply statistics/table.go]: `EstimateTypeWidth` returns memory size in `ListInDisk` while `GetAvgRowSize` returns encoded size in TiKV, I think `pseudoColSize` (aka 8.0) is better.

**Comment by eurekaka** [comment]
@SunRunAway please fix CI.

**Pull request merged**